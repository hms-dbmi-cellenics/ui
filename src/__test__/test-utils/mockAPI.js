import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
