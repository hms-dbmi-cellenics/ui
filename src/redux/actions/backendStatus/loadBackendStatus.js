import handleError from 'utils/http/handleError';
import fetchAPI from 'utils/http/fetchAPI';
import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_LOADED,
  BACKEND_STATUS_ERROR,
} from 'redux/actionTypes/backendStatus';
import endUserMessages from 'utils/endUserMessages';

const loadBackendStatus = (experimentId) => async (dispatch) => {
  dispatch({
    type: BACKEND_STATUS_LOADING,
    payload: {
      experimentId,
    },
  });

  const url = `/v1/experiments/${experimentId}/backendStatus`;
  try {
    const status = await fetchAPI(url);
    dispatch({
      type: BACKEND_STATUS_LOADED,
      payload: {
        experimentId,
        status,
      },
    });

    return status;
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_BACKEND_STATUS);

    dispatch({
      type: BACKEND_STATUS_ERROR,
      payload: {
        experimentId,
        error: errorMessage,
      },
    });
  }
};

export default loadBackendStatus;
