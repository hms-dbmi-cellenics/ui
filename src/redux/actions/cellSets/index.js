import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import deleteCellClass from './deleteCellClass';
import runCellSetsClustering from './runCellSetsClustering';
import runCellSetsAnnotation from './runCellSetsAnnotation';
import runCellCycleScoring from './runCellCycleScoring';
import updateCellSetsClustering from './updateCellSetsClustering';

import loadCellSets from './loadCellSets';
import reorderCellSet from './reorderCellSet';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  deleteCellClass,
  runCellSetsClustering,
  runCellSetsAnnotation,
  updateCellSetsClustering,
  loadCellSets,
  reorderCellSet,
  updateCellSetSelected,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
  runCellCycleScoring,
};
