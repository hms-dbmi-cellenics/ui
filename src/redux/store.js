/* eslint-disable global-require */
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { createWrapper } from 'next-redux-wrapper';
import { enableMapSet } from 'immer';

import rootReducer from 'redux/reducers/index';

enableMapSet();

const bindMiddleware = (middleware) => {
  const { composeWithDevTools } = require('redux-devtools-extension');

  // eslint-disable-next-line import/no-extraneous-dependencies
  const { createLogger } = require('redux-logger');

  // do not log server-side redux actions
  middleware.push(createLogger({
    predicate: () => typeof window !== 'undefined',
  }));

  return composeWithDevTools(applyMiddleware(...middleware));
};

const makeStore = () => {
  const store = createStore(
    rootReducer,
    bindMiddleware([thunk]),
  );

  // IF REDUCERS WERE CHANGED, RELOAD WITH INITIAL STATE
  if (module.hot) {
    module.hot.accept('./reducers', () => {
      const createNextReducer = require('./reducers').default;
      store.replaceReducer(createNextReducer(rootReducer));
    });
  }
  return store;
};

const wrapper = createWrapper(makeStore, { debug: false });
export { wrapper, makeStore };
