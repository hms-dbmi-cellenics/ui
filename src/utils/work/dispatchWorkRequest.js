import dayjs from 'dayjs';
import { Environment } from 'utils/deploymentInfo';

import fetchAPI from 'utils/http/fetchAPI';
// Disable unique keys to reallow reuse of work results in development
const DISABLE_UNIQUE_KEYS = [
  'GetEmbedding',
];

const getCacheUniquenessKey = (taskName, environment) => {
  // Disable cache in development or if localStorage says so
  // Do not disable for embeddings requests because download seurat & trajectory depend on that ETag
  if (
    environment !== Environment.PRODUCTION
    && (localStorage.getItem('disableCache') === 'true' || environment === Environment.DEVELOPMENT)
    && !DISABLE_UNIQUE_KEYS.includes(taskName)
  ) {
    return Math.random();
  }

  return null;
};

const getWorkerTimeout = (taskName, defaultTimeout) => {
  switch (taskName) {
    case 'GetEmbedding':
    case 'ListGenes':
    case 'MarkerHeatmap': {
      return dayjs().add(1800, 's').toISOString();
    }

    default: {
      return dayjs().add(defaultTimeout, 's').toISOString();
    }
  }
};

const dispatchWorkRequest = async (
  experimentId,
  body,
  timeout,
  requestProps,
  getState,
) => {
  const { name: taskName } = body;

  // for listGenes, markerHeatmap, & getEmbedding we set a long timeout for the worker
  // after that timeout the worker will skip those requests
  // meanwhile in the UI we set a shorter timeout. The UI will be prolonging this timeout
  // as long as it receives "heartbeats" from the worker because that means the worker is alive
  // and progresing.
  // this should be removed if we make each request run in a different worker
  const workerTimeoutDate = getWorkerTimeout(taskName, timeout);

  const { environment } = getState().networkResources;
  const cacheUniquenessKey = getCacheUniquenessKey(taskName, environment);

  // eslint-disable-next-line no-param-reassign
  requestProps.cacheUniquenessKey = cacheUniquenessKey;

  const request = {
    experimentId,
    timeout: workerTimeoutDate,
    body,
    requestProps,

  };

  const { data: { ETag, signedUrl } } = await fetchAPI(
    `/v2/workRequest/${experimentId}`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    },
  );

  return { ETag, signedUrl, request };
};

export default dispatchWorkRequest;
