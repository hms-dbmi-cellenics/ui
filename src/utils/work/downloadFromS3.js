import unpackResult from 'utils/work/unpackResult';
import httpStatusCodes from 'utils/http/httpStatusCodes';
import parseResult from 'utils/work/parseResult';

const downloadFromS3 = async (taskName, signedUrl) => {
  const response = await fetch(signedUrl);

  // some WorkRequests like scType and runClustering do not upload data to S3
  // (nor return it via socket) instead they patch the cellsets through the API.
  // In those cases the workResults will not exist, and it's fine because data will
  // be updated through the cellsets.
  // the forbidden (in addition to the not found) is required because of how signed URLs work
  //  when you try to download a file from a signedUrl that doesn't exist you get a 403 forbidden
  // because the user is not authorized to access a file that does not exist
  if (response.status === httpStatusCodes.NOT_FOUND
    || response.status === httpStatusCodes.FORBIDDEN) {
    return null;
  }
  if (!response.ok) {
    throw new Error(`Error ${response.status}: ${response.text}`, { cause: response });
  }

  const unpackedResult = await unpackResult(response, taskName);
  const parsedResult = parseResult(unpackedResult, taskName);

  return parsedResult;
};

export default downloadFromS3;
