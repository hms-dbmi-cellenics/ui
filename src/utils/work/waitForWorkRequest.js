import dayjs from 'dayjs';

import { decompressUint8Array } from 'utils/work/unpackResult';
import parseResult from 'utils/work/parseResult';
import WorkTimeoutError from 'utils/errors/http/WorkTimeoutError';
import WorkResponseError from 'utils/errors/http/WorkResponseError';

import { updateBackendStatus } from 'redux/actions/backendStatus';

const timeoutIds = {};

// getTimeoutDate returns the date resulting of adding 'timeout' seconds to
// current time.
const getTimeoutDate = (timeout) => dayjs().add(timeout, 's').toISOString();

// set a timeout and save its ID in the timeoutIds map
// if there was a timeout for the current ETag, clear it before reseting it
const setOrRefreshTimeout = (request, timeoutDuration, reject, ETag) => {
  if (timeoutIds[ETag]) {
    clearTimeout(timeoutIds[ETag]);
  }

  // timeoutDate needs to be initialized outside the timeout callback
  // or it will be computed when the error is raised instead of (now)
  // when the timeout is set
  const timeoutDate = getTimeoutDate(timeoutDuration);
  timeoutIds[ETag] = setTimeout(() => {
    reject(new WorkTimeoutError(timeoutDuration, timeoutDate, request, ETag));
  }, timeoutDuration * 1000);
};

const getWorkerTimeout = (taskName, defaultTimeout) => {
  switch (taskName) {
    case 'GetEmbedding':
    case 'ListGenes':
    case 'MarkerHeatmap':
    case 'GetNormalizedExpression': {
      return dayjs().add(1800, 's').toISOString();
    }

    default: {
      return dayjs().add(defaultTimeout, 's').toISOString();
    }
  }
};

const waitForWorkRequest = async (
  ETag,
  experimentId,
  request,
  timeout,
  dispatch,
) => {
  const { default: connectionPromise } = await import('utils/socketConnection');

  const io = await connectionPromise;

  // for listGenes, markerHeatmap, & getEmbedding we set a long timeout for the worker
  // after that timeout the worker will skip those requests
  // meanwhile in the UI we set a shorter timeout. The UI will be prolonging this timeout
  // as long as it receives "heartbeats" from the worker because that means the worker is alive
  // and progresing.
  // this should be removed if we make each request run in a different worker
  const workerTimeoutDate = getWorkerTimeout(request, timeout);

  const timeoutPromise = new Promise((resolve, reject) => {
    setOrRefreshTimeout(request, timeout, reject, ETag);

    // this experiment update is received whenever a worker finishes any work request
    // related to the current experiment. We extend the timeout because we know
    // the worker is alive and was working on another request of our experiment
    io.on(`Heartbeat-${experimentId}`, (message) => {
      const newTimeoutDate = getTimeoutDate(timeout);
      if (newTimeoutDate < workerTimeoutDate) {
        const status = {
          worker: {
            statusCode: message.status_code,
            userMessage: message.user_message,
          },
        };
        dispatch(updateBackendStatus(experimentId, status));

        setOrRefreshTimeout(request, timeout, reject, ETag);
      }
    });
  });

  const responsePromise = new Promise((resolve, reject) => {
    io.on(`WorkResponse-${ETag}`, async (res) => {
      // If type is object, then we received a notification with a signedUrl
      // now we need to fetch the actual result from s3
      if (typeof res === 'object') {
        const { response } = res;

        if (response.error) {
          const { errorCode, userMessage } = response;

          return reject(
            new WorkResponseError(errorCode, userMessage, request),
          );
        }

        return resolve({ signedUrl: response.signedUrl });
      }
      // If type isn't object, then we have the actual work result,
      // no further downloads are necessary, we just need to decompress and return it
      const decompressedData = await decompressUint8Array(Uint8Array.from(Buffer.from(res, 'base64')));

      return resolve({ data: parseResult(decompressedData) });
    });
  });

  // TODO switch to using normal WorkRequest for v2 requests
  return await Promise.race([timeoutPromise, responsePromise]);
};

export default waitForWorkRequest;
