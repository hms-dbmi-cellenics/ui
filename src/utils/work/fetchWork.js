import { Environment, isBrowser } from 'utils/deploymentInfo';

import cache from 'utils/cache';
import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';
import downloadFromS3 from 'utils/work/downloadFromS3';
import waitForWorkRequest from 'utils/work/waitForWorkRequest';

// retrieveData will try to get the data for the given experimentId and ETag from
// the fastest source possible. It will try to get the data in order from:
// 1. Browser cache
// 2. S3 (cache) via signedURL
// 3. Worker
//   3.1. Via socket (small data)
//   3.2. Via S3 (large data) via signedURL
const retrieveData = async (experimentId,
  ETag,
  signedUrl,
  request,
  timeout,
  body,
  dispatch,
  useCache) => {
  // 1. Check if we have the ETag in the browser cache (no worker)
  const cachedData = useCache ? await cache.get(ETag) : null;
  if (cachedData) {
    return cachedData;
  }

  // 2. Check if data is cached in S3 so we can download from the signed URL (no worker)
  if (signedUrl) {
    return await downloadFromS3(body.name, signedUrl);
  }

  // 3. If we don't have signedURL, wait for the worker to send us the data via
  // - the data via socket
  // - the signedURL to download the data from S3
  const { signedUrl: workerSignedUrl, data } = await waitForWorkRequest(
    ETag,
    experimentId,
    request,
    timeout,
    dispatch,
  );

  // 3.1. The worker send the data via socket because it's small enough
  if (data) {
    return data;
  }
  // 3.2. The worker send a signedUrl to download the data
  return await downloadFromS3(body.name, workerSignedUrl);
};

const fetchWork = async (
  experimentId,
  body,
  getState,
  dispatch,
  optionals = {},
) => {
  const {
    extras = undefined,
    timeout = 180,
    broadcast = false,
    onETagGenerated = () => { },
  } = optionals;

  if (!isBrowser) {
    throw new Error('Disabling network interaction on server');
  }

  const { environment } = getState().networkResources;
  if (environment === Environment.DEVELOPMENT && !localStorage.getItem('disableCache')) {
    localStorage.setItem('disableCache', 'true');
  }

  // 1. Contact the API to get ETag and possible S3 signedURL
  const { ETag, signedUrl, request } = await dispatchWorkRequest(
    experimentId,
    body,
    timeout,
    {
      broadcast,
      ...extras,
    },
    dispatch,
  );

  onETagGenerated(ETag);

  const useCache = body.name !== 'GeneExpression';

  // 2. Try to get the data from the fastest source possible
  const data = await retrieveData(
    experimentId, ETag, signedUrl, request, timeout, body, dispatch, useCache,
  );

  // 3. Cache the data in the browser
  if (useCache) {
    await cache.set(ETag, data);
  }

  return data;
};

export default fetchWork;
