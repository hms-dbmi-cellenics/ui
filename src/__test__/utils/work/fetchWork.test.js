/* eslint-disable global-require */
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import mockAPI, { generateDefaultMockAPIResponses } from '__test__/test-utils/mockAPI';
import fake from '__test__/test-utils/constants';

import fetchWork from 'utils/work/fetchWork';
import { loadBackendStatus } from 'redux/actions/backendStatus';
import { makeStore } from 'redux/store';
import downloadFromS3 from 'utils/work/downloadFromS3';
import waitForWorkRequest from 'utils/work/waitForWorkRequest';
import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';

const {
  mockCacheGet,
  mockCacheSet,
  mockSeekFromS3,
} = require('__test__/utils/work/fetchWork.mock');

jest.mock(
  'utils/cache',
  () => require('__test__/utils/work/fetchWork.mock').mockCacheModule,
);

jest.mock('utils/work/dispatchWorkRequest', () => jest.fn().mockReturnValue({
  ETag,
  signedUrl: 'fakeSignedUrl',
  request: null,
}));

jest.mock('utils/work/downloadFromS3');
jest.mock('utils/work/waitForWorkRequest');

const timeout = 10;
const ETag = 'fakeETag';
const nonGeneExpressionWorkRequest = {
  name: 'ListGenes',
};

const geneExpressionWorkRequest = {
  name: 'GeneExpression',
  genes: ['A', 'B', 'C', 'D'],
};

enableFetchMocks();

describe('fetchWork', () => {
  let store;

  beforeEach(async () => {
    Storage.prototype.setItem = jest.fn();

    jest.clearAllMocks();

    fetchMock.resetMocks();
    fetchMock.mockIf(/.*/, mockAPI(generateDefaultMockAPIResponses(fake.EXPERIMENT_ID)));

    store = makeStore();

    await store.dispatch(loadBackendStatus(fake.EXPERIMENT_ID));
  });

  it('returns data from cache if available', async () => {
    mockCacheGet.mockImplementationOnce(() => ({ cacheData: true }));

    const res = await fetchWork(
      fake.EXPERIMENT_ID,
      nonGeneExpressionWorkRequest,
      store.getState,
      store.dispatch,
      { timeout },
    );

    expect(mockCacheGet).toHaveBeenCalledWith(ETag);
    expect(downloadFromS3).not.toHaveBeenCalled();
    expect(waitForWorkRequest).not.toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('returns data from S3 directly if available', async () => {
    downloadFromS3.mockReturnValueOnce({
      S3Data: true,
    });

    const res = await fetchWork(
      fake.EXPERIMENT_ID,
      nonGeneExpressionWorkRequest,
      store.getState,
      store.dispatch,
      { timeout },
    );

    expect(mockCacheGet).toHaveBeenCalledWith(ETag);
    expect(downloadFromS3.mock.calls).toMatchSnapshot();
    expect(waitForWorkRequest).not.toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('waits and returns data from the worker', async () => {
    dispatchWorkRequest.mockReturnValueOnce({
      ETag,
      signedUrl: null,
      request: null,
    });

    waitForWorkRequest.mockReturnValueOnce({
      workerSignedUrl: 'fakeWorkerSignedUrl',
      data: true,
    });

    const res = await fetchWork(
      fake.EXPERIMENT_ID,
      nonGeneExpressionWorkRequest,
      store.getState,
      store.dispatch,
      { timeout },
    );

    expect(mockCacheGet).toHaveBeenCalledWith(ETag);
    expect(downloadFromS3.mock.calls).toMatchSnapshot();
    expect(waitForWorkRequest.mock.calls).toMatchSnapshot();
    expect(res).toMatchSnapshot();
  });

  it('does not use cache for gene expression request', async () => {
    downloadFromS3.mockReturnValueOnce({
      S3Data: true,
    });

    const res = await fetchWork(
      fake.EXPERIMENT_ID,
      geneExpressionWorkRequest,
      store.getState,
      store.dispatch,
      { timeout },
    );

    expect(mockCacheGet).not.toHaveBeenCalled();
    expect(mockCacheSet).not.toHaveBeenCalled();
    expect(downloadFromS3.calls).toMatchSnapshot();
    expect(waitForWorkRequest).not.toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('Throws an error if the dispatched work request throws an error', async () => {
    dispatchWorkRequest
      .mockReset()
      .mockImplementationOnce(() => Promise.reject(new Error('Worker timeout')));

    await expect(
      fetchWork(
        fake.EXPERIMENT_ID,
        nonGeneExpressionWorkRequest,
        store.getState,
        store.dispatch,
        { timeout: 10 },
      ),
    ).rejects.toThrow();

    expect(mockCacheGet).not.toHaveBeenCalled();
    expect(mockCacheSet).not.toHaveBeenCalled();

    // Not called ever because result is received straight from dispatchWorkRequest
    expect(mockSeekFromS3).toHaveBeenCalledTimes(0);
  });
});
