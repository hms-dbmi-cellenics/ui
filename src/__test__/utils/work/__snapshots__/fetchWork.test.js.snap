// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchWork does not use cache for gene expression request 1`] = `undefined`;

exports[`fetchWork does not use cache for gene expression request 2`] = `
{
  "S3Data": true,
}
`;

exports[`fetchWork returns data from S3 directly if available 1`] = `
[
  [
    "ListGenes",
    "fakeSignedUrl",
  ],
]
`;

exports[`fetchWork returns data from S3 directly if available 2`] = `
{
  "S3Data": true,
}
`;

exports[`fetchWork returns data from cache if available 1`] = `
{
  "cacheData": true,
}
`;

exports[`fetchWork waits and returns data from the worker 1`] = `[]`;

exports[`fetchWork waits and returns data from the worker 2`] = `
[
  [
    "fakeETag",
    "testae48e318dab9a1bd0bexperiment",
    null,
    10,
    [Function],
  ],
]
`;

exports[`fetchWork waits and returns data from the worker 3`] = `true`;
